File: book/lib/services/file_service.dart
import 'api_client.dart';

class FileItem {
  final String name;
  final String path;
  final bool isDirectory;
  final int size;
  final DateTime modifiedAt;

  FileItem({
    required this.name,
    required this.path,
    required this.isDirectory,
    required this.size,
    required this.modifiedAt,
  });

  factory FileItem.fromJson(Map<String, dynamic> json) {
    return FileItem(
      name: json['name'] as String? ?? '',
      path: json['path'] as String? ?? '',
      isDirectory: json['is_directory'] as bool? ??
                   json['file_type'] == 'directory' ||
                   json['type'] == 'directory',
      size: json['size'] as int? ?? 0,
      modifiedAt: json['modified_at'] != null
          ? DateTime.parse(json['modified_at'] as String)
          : DateTime.now(),
    );
  }
}

class StorageInfo {
  final int used;
  final int total;

  StorageInfo({required this.used, required this.total});

  factory StorageInfo.fromJson(Map<String, dynamic> json) {
    return StorageInfo(
      used: json['used'] as int? ?? 0,
      total: json['total'] as int? ?? 0,
    );
  }

  double get usedGB => used / (1024 * 1024 * 1024);
  double get totalGB => total / (1024 * 1024 * 1024);
  double get percent => total > 0 ? (used / total) * 100 : 0;
}

class FileService {
  final ApiClient _api;

  FileService({ApiClient? api}) : _api = api ?? apiClient;

  Future<List<FileItem>> list(String path) async {
    try {
      final response = await _api.get('/api/files?path=$path');
      final files = response['files'] as List<dynamic>? ?? [];
      return files.map((f) => FileItem.fromJson(f as Map<String, dynamic>)).toList();
    } catch (e) {
      return [];
    }
  }

  Future<StorageInfo> getStorageInfo() async {
    try {
      final response = await _api.get('/api/files/storage');
      return StorageInfo.fromJson(response);
    } catch (e) {
      return StorageInfo(used: 0, total: 0);
    }
  }

  Future<bool> createDirectory(String path) async {
    try {
      await _api.post('/api/files/mkdir', {'path': path});
      return true;
    } catch (e) {
      return false;
    }
  }

  Future<bool> delete(String path) async {
    try {
      await _api.delete('/api/files?path=$path');
      return true;
    } catch (e) {
      return false;
    }
  }

  Future<String?> read(String path) async {
    try {
      final response = await _api.get('/api/files/read?path=$path');
      return response['content'] as String?;
    } catch (e) {
      return null;
    }
  }

  Future<bool> uploadFile(String filename, List<int> bytes, String path) async {
    try {
      await _api.uploadFile('/api/files/upload', filename, bytes, path);
      return true;
    } catch (e) {
      return false;
    }
  }
}

final fileService = FileService();

File: notebook/src/services/fileService.ts
// File Service - File management operations

import apiClient from "./apiClient";

export interface FileInfo {
  name: string;
  path: string;
  isDirectory: boolean;
  size: number;
  modifiedAt: Date;
  createdAt: Date;
  mimeType?: string;
  extension?: string;
}

export interface StorageInfo {
  usedBytes: number;
  totalBytes: number;
  freeBytes: number;
  usedPercent: number;
}

export interface FileContent {
  path: string;
  content: string;
  encoding: string;
  mimeType: string;
}

export interface DatasetPreview {
  path: string;
  format: string;
  columns: string[];
  rows: Array<Record<string, unknown>>;
  totalRows: number;
  schema: Array<{
    name: string;
    type: string;
    nullable: boolean;
  }>;
}

interface FileInfoResponse {
  name: string;
  path: string;
  file_type: "file" | "directory";
  size: number;
  modified_at: string;
  mime_type?: string | null;
}

interface FileListResponse {
  path: string;
  files: FileInfoResponse[];
}

class FileService {
  private parseFileInfo(data: FileInfoResponse): FileInfo {
    return {
      name: data.name,
      path: data.path,
      isDirectory: data.file_type === "directory",
      size: data.size,
      modifiedAt: new Date(data.modified_at),
      createdAt: new Date(data.modified_at), // Backend doesn't provide created_at
      mimeType: data.mime_type || undefined,
      extension: data.name.includes(".") ? data.name.split(".").pop() : undefined,
    };
  }

  async list(path: string = ""): Promise<FileInfo[]> {
    const response = await apiClient.get<FileListResponse>(`/api/files${path ? `?path=${encodeURIComponent(path)}` : ""}`);
    return (response.files || []).map((f) => this.parseFileInfo(f));
  }

  async getInfo(path: string): Promise<FileInfo> {
    const response = await apiClient.get<FileInfoResponse>(`/api/files/info?path=${encodeURIComponent(path)}`);
    return this.parseFileInfo(response);
  }

  async createDirectory(path: string): Promise<FileInfo> {
    const response = await apiClient.post<FileInfoResponse>("/api/files/directory", { path });
    return this.parseFileInfo(response);
  }

  async upload(file: File, destinationPath: string): Promise<FileInfo> {
    const response = await apiClient.uploadFile<FileInfoResponse>("/api/files/upload", file, {
      path: destinationPath,
    });
    return this.parseFileInfo(response);
  }

  async uploadMultiple(files: File[], destinationPath: string): Promise<FileInfo[]> {
    const results: FileInfo[] = [];
    for (const file of files) {
      const result = await this.upload(file, destinationPath);
      results.push(result);
    }
    return results;
  }

  async delete(path: string): Promise<void> {
    await apiClient.delete(`/api/files?path=${encodeURIComponent(path)}`);
  }

  async deleteMultiple(paths: string[]): Promise<void> {
    await apiClient.post("/api/files/delete-multiple", { paths });
  }

  async rename(oldPath: string, newPath: string): Promise<FileInfo> {
    const response = await apiClient.post<FileInfoResponse>("/api/files/rename", {
      old_path: oldPath,
      new_path: newPath,
    });
    return this.parseFileInfo(response);
  }

  async move(sourcePath: string, destinationPath: string): Promise<FileInfo> {
    const response = await apiClient.post<FileInfoResponse>("/api/files/move", {
      source_path: sourcePath,
      destination_path: destinationPath,
    });
    return this.parseFileInfo(response);
  }

  async copy(sourcePath: string, destinationPath: string): Promise<FileInfo> {
    const response = await apiClient.post<FileInfoResponse>("/api/files/copy", {
      source_path: sourcePath,
      destination_path: destinationPath,
    });
    return this.parseFileInfo(response);
  }

  async read(path: string): Promise<FileContent> {
    const response = await apiClient.get<{
      path: string;
      content: string;
      encoding: string;
      mime_type: string;
    }>(`/api/files/read?path=${encodeURIComponent(path)}`);

    return {
      path: response.path,
      content: response.content,
      encoding: response.encoding,
      mimeType: response.mime_type,
    };
  }

  async write(path: string, content: string): Promise<FileInfo> {
    const response = await apiClient.post<FileInfoResponse>("/api/files/write", {
      path,
      content,
    });
    return this.parseFileInfo(response);
  }

  async getDownloadUrl(path: string): Promise<string> {
    return `${apiClient.getBaseUrl()}/api/files/download?path=${encodeURIComponent(path)}`;
  }

  async download(path: string): Promise<Blob> {
    const url = await this.getDownloadUrl(path);
    const response = await fetch(url);
    return response.blob();
  }

  async getStorageInfo(): Promise<StorageInfo> {
    const response = await apiClient.get<{
      used_bytes: number;
      total_bytes: number;
      free_bytes: number;
      used_percent: number;
    }>("/api/files/storage");

    return {
      usedBytes: response.used_bytes,
      totalBytes: response.total_bytes,
      freeBytes: response.free_bytes,
      usedPercent: response.used_percent,
    };
  }

  // Dataset operations
  async previewDataset(path: string, limit: number = 100): Promise<DatasetPreview> {
    const response = await apiClient.get<{
      path: string;
      format: string;
      columns: string[];
      rows: Array<Record<string, unknown>>;
      total_rows: number;
      schema: Array<{
        name: string;
        type: string;
        nullable: boolean;
      }>;
    }>(`/api/files/preview?path=${encodeURIComponent(path)}&limit=${limit}`);

    return {
      path: response.path,
      format: response.format,
      columns: response.columns,
      rows: response.rows,
      totalRows: response.total_rows,
      schema: response.schema,
    };
  }

  // Search files
  async search(query: string, path: string = "/"): Promise<FileInfo[]> {
    const response = await apiClient.get<FileInfoResponse[]>(
      `/api/files/search?query=${encodeURIComponent(query)}&path=${encodeURIComponent(path)}`
    );
    return response.map((f) => this.parseFileInfo(f));
  }

  // Get file type icon based on extension
  getFileType(filename: string): "python" | "image" | "data" | "notebook" | "text" | "other" {
    const ext = filename.split(".").pop()?.toLowerCase() || "";

    if (ext === "py") return "python";
    if (["png", "jpg", "jpeg", "gif", "webp", "svg", "bmp"].includes(ext)) return "image";
    if (["csv", "json", "xlsx", "xls", "parquet", "feather", "pickle", "pkl"].includes(ext)) return "data";
    if (ext === "ipynb") return "notebook";
    if (["txt", "md", "rst", "yaml", "yml", "toml", "ini", "cfg", "log"].includes(ext)) return "text";
    return "other";
  }

  // Format file size
  formatSize(bytes: number): string {
    if (bytes === 0) return "0 B";
    const k = 1024;
    const sizes = ["B", "KB", "MB", "GB", "TB"];
    const i = Math.floor(Math.log(bytes) / Math.log(k));
    return `${parseFloat((bytes / Math.pow(k, i)).toFixed(1))} ${sizes[i]}`;
  }
}

export const fileService = new FileService();
export default fileService;
